### PR TITLE
Move loading of .hsgr to io.hpp

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

This is the first part of https://github.com/Project-OSRM/osrm-backend/issues/2989

## Tasklist
 - [x] review
 - [x] adjust for for comments

/cc @karenzshea @chaupow @miccolis @danpat

### CODE REVIEW

File: include/storage/io.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace osrm
+{
+namespace storage
+{
+namespace io
+{
+
+// Returns the checksum and the number of nodes and number of edges
+std::tuple<unsigned, std::size_t, std::size_t> readHSGRSize(boost::filesystem::ifstream &input_stream)
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: Maybe call it `readHSGRMetadata` (?) - also all of those functions will return a fingerprint and some additional data, right? What about making a namedtuple type instead so call sites don't have to `get<Nth>` all over the place.

    struct { Fingerprint fingerprint; size_t nbNodes; size_t nbEdges; }

or similar.

File: include/storage/io.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+{
+    const util::FingerPrint fingerprint_valid = util::FingerPrint::GetValid();
+    util::FingerPrint fingerprint_loaded;
+    input_stream.read(reinterpret_cast<char *>(&fingerprint_loaded), sizeof(util::FingerPrint));
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: read can fail after which the stream's bool conversion will fail - `if (!stream) ..`

File: include/storage/io.hpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    if (!fingerprint_loaded.TestGraphUtil(fingerprint_valid))
+    {
+        util::SimpleLogger().Write(logWARNING) << ".hsgr was prepared with different build.\n"
+                                                  "Reprocess to get rid of this warning.";
+    }
+
+    unsigned check_sum = 0;
+    unsigned number_of_nodes = 0;
+    unsigned number_of_edges = 0;
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: Same here: fixed-size for all of those? (Also at write-time then)

File: include/storage/io.hpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    return std::make_tuple(check_sum, number_of_nodes, number_of_edges);
+}
+
+// Needs to be called after getHSGRSize() to get the correct offset in the stream
+template <typename NodeT, typename EdgeT>
+void readHSGR(boost::filesystem::ifstream &input_stream,
+              NodeT *node_buffer,
+              std::size_t number_of_nodes,
+              EdgeT *edge_buffer,
+              std::size_t number_of_edges)
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: Can those raw pointers ever be null?

File: include/storage/io.hpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    BOOST_ASSERT_MSG(0 != number_of_nodes, "number of nodes is zero");
+    BOOST_ASSERT_MSG(0 != number_of_edges, "number of edges is zero");
+    input_stream.read(reinterpret_cast<char *>(check_sum), sizeof(unsigned));
+    input_stream.read(reinterpret_cast<char *>(&number_of_nodes), sizeof(unsigned));
+    input_stream.read(reinterpret_cast<char *>(&number_of_edges), sizeof(unsigned));
+
+    return std::make_tuple(check_sum, number_of_nodes, number_of_edges);
+}
+
+// Needs to be called after getHSGRSize() to get the correct offset in the stream
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: Why do I have to care about this? if I _have to_ do something, why don't you already do it for me?

File: src/storage/storage.cpp [hunk 6]
```diff
@@ -189,35 +190,13 @@ int Storage::Run()
...
-        util::SimpleLogger().Write(logWARNING) << ".hsgr was prepared with different build. "
-                                                  "Reprocess to get rid of this warning.";
-    }
-
-    // load checksum
     unsigned checksum = 0;
-    hsgr_input_stream.read((char *)&checksum, sizeof(unsigned));
+    std::size_t number_of_graph_nodes = 0;
+    std::size_t number_of_graph_edges = 0;
+    std::tie(checksum, number_of_graph_nodes, number_of_graph_edges) = io::readHSGRSize(hsgr_input_stream);
```
Comment (daniel-j-h) [inline src/storage/storage.cpp]: Depends on tuple order - there are no named product types in C++ (think Python's namedtuple) but at least create a small type and use semantic member names here.

File: include/storage/io.hpp (hunk 2 above)
Comment (TheMarex) [inline include/storage/io.hpp]: Can we make it throw an exception always?

File: include/storage/io.hpp (hunk 4 above)
Comment (MoKob) [inline include/storage/io.hpp]: Also, is there a reason that we have to use raw pointers at all?

**Comment by TheMarex** [comment]
@daniel-j-h please confirm if this works for you now.

File: include/storage/io.hpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+// 
+template <typename NodeT, typename EdgeT>
+void readHSGR(boost::filesystem::ifstream &input_stream,
+              NodeT *node_buffer,
+              std::uint32_t number_of_nodes,
+              EdgeT *edge_buffer,
+              std::uint32_t number_of_edges)
+{
+    input_stream.read(reinterpret_cast<char *>(node_buffer), number_of_nodes * sizeof(NodeT));
+    input_stream.read(reinterpret_cast<char *>(edge_buffer), number_of_edges * sizeof(EdgeT));
```
Comment (daniel-j-h) [inline include/storage/io.hpp]: These can still fail leaving the stream corrupted - the call site does not check for this either. But it's a more general problem of this code base..

**Pull request merged**